Tasks/Maven/maventask.ts

```diff
@@ -303,70 +333,110 @@ function cleanDirectory(targetDirectory:string):Q.Promise<void> {
-        var analysisResults:ar.AnalysisResult[] = [];
-
-        // PMD
-        if (tl.getInput('pmdAnalysisEnabled', true) == 'true') {
-            try {
-                var pmdResults:ar.AnalysisResult = pmd.processPmdOutput(sourcesDir);
-                if (pmdResults) {
-                    analysisResults.push(pmdResults);
+    var enabledCodeAnalysisTools = getEnabledCodeAnalysisTools();
+    if (enabledCodeAnalysisTools.length > 0) {
+        // Discover maven modules
+        var modules:ma.ModuleAnalysis[] = findModules(sourcesDir);
+
+        // Special case: if the root turns up as a module, the automatic name won't do
+        modules.forEach((module) => {
+            if (module.rootFolder == sourcesDir) {
+                module.moduleName = 'root'
+            }
+        });
+
+        tl.debug('Discovered ' + modules.length + ' Maven modules to upload results from.');
+
+        // Discover output files, summarise per-module results
+        modules.forEach((module) => {
+            // PMD
+            if (enabledCodeAnalysisTools.indexOf(pmd.toolName) > -1) {
+                try {
+                    var pmdResults:ar.AnalysisResult = pmd.processPmdOutput(module.rootFolder);
+                    if (pmdResults) {
+                        module.analysisResults[pmdResults.toolName] = pmdResults;
+                    }
+                } catch(e) {
+                    tl.error(e.message);
+                    tl.exit(1);
                 }
-            } catch(e) {
-                tl.error(e.message);
-                tl.exit(1);
             }
-        }
 
-        // Process analysis results - upload files and generate build summary lines
+        });
+
 
-        var buildSummaryString:string = '';
+        // Gather analysis results by tool for summaries, stage and upload build artifacts
+        var buildSummaryLines:string[] = [];
         cleanDirectory(stagingDir).then(() => {
-            analysisResults.forEach(analysisResult => {
-                var buildSummaryLine:string = buildSummaryLineFromResult(analysisResult);
-                buildSummaryString += buildSummaryLine + "  \r\n";
-                uploadBuildArtifactsFromResult(analysisResult);
+            enabledCodeAnalysisTools.forEach((toolName:string) => {
+                var toolAnalysisResults:ar.AnalysisResult[] = [];
+                modules.forEach((module:ma.ModuleAnalysis) => {
+                    var moduleAnalysisResult:ar.AnalysisResult = module.analysisResults[toolName];
+                    if (moduleAnalysisResult) {
+                        toolAnalysisResults.push(moduleAnalysisResult);
+                    }
+
+                    uploadBuildArtifacts(toolName, module);
+                });
+
+                // After looping through all modules, summarise tool output results
+                var summaryLine:string = createSummaryLine(toolName, toolAnalysisResults);
+                buildSummaryLines.push(summaryLine);
```

Comment:
While the gathering of files is more easily done per-module, the summary of results is done per-tool. Thus the second loop iterates over the enabled code analysis tools.

An inner for-each iterates through the module list. If the given module contained an AnalysisResult object for the current tool, we store the result and continue on to upload an artifact on a per-module, per-tool basis. 

After this, the summary line for the tool is created by iterating through all the AnalysisResult instances for the current tool and summing the total violations and files with violations.